segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageProvider.java

```diff
@@ -0,0 +1,178 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.chunklayer;
+
+import java.io.InputStream;
+
+/**
+ * Defines an abstraction for Permanent Storage.
+ * Note: not all operations defined here are needed.
+ * <div>
+ * Below are minimum requirements that any implementation must provide.
+ * Note that it is the responsibility of storage provider specific implementation to make sure following guarantees are provided even
+ * though underlying storage may not provide all primitives or guarantees.
+ * <ul>
+ * <li>Once an operation is executed and acknowledged as successful then the effects must be permanent and consistent (as opposed to eventually consistent)</li>
+ * <li>{@link ChunkStorageProvider#create(String)}  and {@link ChunkStorageProvider#delete(ChunkHandle)} are not idempotent.</li>
+ * <li>{@link ChunkStorageProvider#exists(String)} and {@link ChunkStorageProvider#getInfo(String)} must reflect effects of most recent operation performed.</li>
+ * </ul>
+ * </div>
+ * <div>
+ * There are a few different capabilities that ChunkStorageProvider needs to provide.
+ * <ul>
+ * <li> Does {@link ChunkStorageProvider}  support appending to existing chunks? For vanilla S3 compatible this would return false. This is indicated by {@link ChunkStorageProvider#supportsAppend()}. </li>
+ * <li> Does {@link ChunkStorageProvider}  support for concatenating chunks natively? This is indicated by {@link ChunkStorageProvider#supportsConcat()}.
+ * If this is true then native concat operation concat will be invoked otherwise concatWithAppend is invoked.</li>
+ * <li>In addition {@link ChunkStorageProvider} may provide ability to truncate chunks at given offsets (either at front end or at tail end). This is indicated by {@link ChunkStorageProvider#supportsTruncation()}. </li>
+ * </ul>
+ * There are some obvious constraints - If ChunkStorageProvider supports concat but not natively then it must support append .
+ *
+ * For concats, {@link ChunkStorageProvider} supports both native and append, ChunkStorageManager will invoke appropriate method depending on size of target and source chunks. (Eg. ECS)
+ * </div>
+ */
+public interface ChunkStorageProvider extends AutoCloseable {
```

Comment:
This class does not sound like a provider, the way I understand is a class related to a factory or factories, while I'm not seeing this at all. Why are we calling it a provider?

Reply:
The usage is somewhat similar to "FileSystemProvider" being the concrete implementation that "provides" the functionality. 
We don't have methods that return Input/output streams but instead have read and write methods. 
https://docs.oracle.com/javase/8/docs/technotes/guides/io/fsp/filesystemprovider.html
https://docs.oracle.com/javase/8/docs/api/java/nio/file/spi/FileSystemProvider.html#newInputStream-java.nio.file.Path-java.nio.file.OpenOption...-

Reply:
Just calling it ChunkStorage is little bit ambiguous. 
Instead I have two long names.  
ChunkStorageManager - manages chunks.
ChunkStorageProvider - provides chunks.

ChunkManager/ChunkStorage might be an alternative naming scheme.